Add support for embedding NAM models and IR files in session data

damiangr opened the pull request:
## Description
This PR adds the ability to embed NAM model files (.nam) and impulse response files (.wav) directly into DAW session/project files, making sessions fully portable and self-contained.

### Features
- **Automatic Embedding**: When saving a session, NAM models and IR files are encoded and embedded in the plugin state
- **Automatic Extraction**: When loading a session, embedded files are extracted to temporary files and loaded automatically
- **Backward Compatible**: Sessions without embedded data continue to work with file paths as before
- **Cross-Platform**: Designed to work on Windows, macOS, and Linux

### Benefits
- Share sessions without worrying about missing model/IR files
- Old projects remain playable even if original files are moved/deleted
- Team members can open sessions without manual file management

### Implementation
- Modified `SerializeState()` to embed file contents in session data
- Modified `UnserializeState()` to detect and extract embedded files automatically
- Added `_StageModelFromData()` and `_StageIRFromData()` helper methods for loading from memory
- Maintain backward compatibility with file path references
- Update AudioDSPTools submodule to include SIGFPE fix (depends on sdatkinson/AudioDSPTools#24)

## PR Checklist
- [x] Did you format your code using [`format.bash`](https://github.com/sdatkinson/NeuralAmpModelerPlugin/blob/main/format.bash)?
- [x] Does the VST3 plugin pass all of the unit tests in the [VST3PluginTestHost](https://steinbergmedia.github.io/vst3_dev_portal/pages/What+is+the+VST+3+SDK/Plug-in+Test+Host.html)?
  - [x] Windows - Validated with pluginval (strictness level 5) - **SUCCESS**
  - [ ] macOS - **Unable to test**: VMware Sonoma 14.7.2 + Xcode 16.2 environment has NanoVG/Metal incompatibility issues. Recommend testing via GitHub Actions CI which uses Xcode 15.x.
- [x] Does your PR add, remove, or rename any plugin parameters? **No**
- [x] Does your PR add or remove any graphical assets? **No**

## Testing Notes
### Windows (Validated ✅)
- Build: Visual Studio 2022, x64 Release
- Validation: pluginval v1.x, strictness level 5
- Results: All tests passed (audio processing, automation, state save/restore, editor, bus configuration)

### macOS (Limited Testing ⚠️)
**Note**: Local macOS testing was not possible due to environmental constraints:
- Test environment: VMware Sonoma 14.7.2 + Xcode 16.2
- Issue: NanoVG/Metal incompatibility causing image loading crashes (not related to this PR)
- Official v0.7.13 release also crashes in this environment
- **Recommendation**: Please test via GitHub Actions CI, which uses Xcode 15.x and should work correctly

The embedding code itself is platform-independent and uses standard C++ file I/O. The implementation should work on macOS, but comprehensive validation via CI is recommended.



## Files Changed
NeuralAmpModeler.cpp: Embedding/extraction logic and helper methods (+260 lines)
NeuralAmpModeler.h: Added mNAMData and mIRData member variables (+8 lines)
Unserialization.cpp: Added embedded data extraction with versioned functions for 0.7.13 (+134 lines)

sdatkinson commented on NeuralAmpModeler/Unserialization.cpp:
```diff
@@ -57,11 +57,47 @@ void NeuralAmpModeler::_UnserializeApplyConfig(nlohmann::json& config)
   mIRPath.Set(static_cast<std::string>(config["IRPath"]).c_str());
 
-  if (mNAMPath.GetLength())
+  // NEW: Prefer embedded data over file paths
```
I think I'd like the opposite, but I'm willing to hear your rationale for this.

sdatkinson commented on NeuralAmpModeler/Unserialization.cpp:
```diff
@@ -57,11 +57,47 @@ void NeuralAmpModeler::_UnserializeApplyConfig(nlohmann::json& config)
   mNAMPath.Set(static_cast<std::string>(config["NAMPath"]).c_str());
   mIRPath.Set(static_cast<std::string>(config["IRPath"]).c_str());
 
```
What's going on here is being done wrong. Since you're defining new logic on how to deserialize for the next version, you should be adding new functions for v0.7.14 similar to how the ones for 0.7.12 were added.

damiangr replied:
created proper _GetConfigFrom_0_7_13() and _UpdateConfigFrom_0_7_13().

sdatkinson reviewed (changes requested):
My overall impression after reading the code is that this is too complicated for what it brings.

The changes below are probably good pieces of advice if you want to use them to improve your fork, but it's sort of surprising how many different places are touched in this PR. Unexpected things may be happening--for example, the sample rate validation.

If that's a problem in the existing code aside from the feature this is trying to introduce, then it'd be worth it to PR those fixes separately (i.e. validate inputs and demonstrate that those validations are failing). But if this is _introducing_ invalid usage of those existing pieces of code, then it's not immediately clear to me how that's happening, and that feels risky for the stability of the plugin.

You definitely want to re-do the unserialization code to follow the existing pattern because that part definitely isn't following the pattern so far and risks breaking existing projects--which would be ironic, given the goal of this PR 🙂

damiangr commented on NeuralAmpModeler/NeuralAmpModeler.h:
```diff
@@ -307,6 +311,11 @@ class NeuralAmpModeler final : public iplug::Plugin
   WDL_String mIRPath;
 
+  // Embedded file data for portability (stored with DAW session)
+  // Mutable so SerializeState (const method) can read files if needed
```
removed.

damiangr commented:
> rows? It feel
Thank you for checking ! 
Regarding the arrows, they will now work as follows:
If the file is not found, but the directory exists, the arrows will work as usual to switch to the next file within the directory.
If the directory does not exist, the arrows do nothing, but the saved embedded files are loaded.

damiangr commented on NeuralAmpModeler/NeuralAmpModeler.cpp:
```diff
@@ -689,14 +770,26 @@ void NeuralAmpModeler::_SetOutputGain()
     std::unique_ptr<nam::DSP> model = nam::get_dsp(dspPath);
-    std::unique_ptr<ResamplingNAM> temp = std::make_unique<ResamplingNAM>(std::move(model), GetSampleRate());
-    temp->Reset(GetSampleRate(), GetBlockSize());
+    std::unique_ptr<ResamplingNAM> temp = std::make_unique<ResamplingNAM>(std::move(model), sampleRate);
```
yes, there was an issue - sample rate could be 0 during unserialization causing division by zero.

damiangr replied:
Double checked and removed - reverted to use [GetSampleRate()] directly like upstream. No issue, this was an unnecessary change.

damiangr committed:
Add support for embedding NAM models and IR files in session data

This enables DAW sessions to be fully portable by storing the actual
NAM model and IR file data in the session, not just file paths.

Changes:
- SerializeState: Embed NAM/IR file data in session chunk
- UnserializeState: Load from embedded data if file path not found
- New _StageModelFromData/_StageIRFromData functions
- Version bump to 0.7.13 with backward-compatible serialization

Behavior:
- Prefers loading from file path if available (for easy model updates)
- Falls back to embedded data if file is missing (for portability)
- Fully backward compatible with older session formats

damiangr commented:
All review feedback addressed:

Removed test-release.yml and prepare_resources-mac.py changes
Removed unnecessary safety checks from _ResetModelAndIR
Reverted _StageModel to use GetSampleRate() directly
Removed mutable - files read at load time now
Created proper versioned functions (_GetConfigFrom_0_7_13, etc.)
Changed to prefer file path over embedded data
Using version 0.7.13
